oxide-vpc/tests/integration_tests.rs

```diff
@@ -1602,6 +1617,238 @@ fn guest_to_internet() {
+        &data[..],
+    );
+
+    assert_eq!(g1.port.stats_snap().out_uft_hit, 0);
+    let res = g1.port.process(Out, &mut pkt3, &mut ameta);
+    assert!(matches!(res, Ok(Modified)), "bad result: {:?}", res);
+    assert_port!(g1);
+
+    // Ether + IPv6 + UDP + Geneve + Ether + IPv4
+    assert_eq!(pkt3.body_offset(), 14 + 40 + 8 + 8 + 14 + 20);
+    assert_eq!(pkt3.body_seg(), 1);
+    let meta = pkt3.meta();
+
+    match meta.inner.ether.as_ref() {
+        Some(eth) => {
+            assert_eq!(eth.src, g1_cfg.private_mac);
+            assert_eq!(eth.dst, g1_cfg.boundary_services.mac);
+            assert_eq!(eth.ether_type, ETHER_TYPE_IPV4);
+        }
+
+        None => panic!("expected inner Ether header"),
+    }
+
+    match meta.inner.ip.as_ref().unwrap() {
+        IpMeta::Ip4(ip4) => {
+            assert_eq!(ip4.src, g1_cfg.snat().external_ip);
+            assert_eq!(ip4.dst, dst_ip);
+            assert_eq!(ip4.proto, Protocol::ICMP);
+        }
+
+        ip6 => panic!("execpted inner IPv4 metadata, got IPv6: {:?}", ip6),
+    }
+
+    let body = pkt3.body_segs().unwrap()[0];
+    let icmp = Icmpv4Packet::new_checked(body).unwrap();
+    assert_eq!(icmp.echo_ident(), mapped_port);
+    assert_eq!(icmp.echo_seq_no(), seq_no);
+    assert_eq!(g1.port.stats_snap().out_uft_hit, 1);
+
+    // ================================================================
+    // Process ICMP Echo Reply a second time. Once again, this time we
+    // want to verify that the body transformation comes from the UFT
+    // entry.
+    // ================================================================
+    let mut pkt4 = gen_icmp_echo_reply(
+        g1_cfg.boundary_services.mac,
+        g1_cfg.private_mac,
+        dst_ip,
+        g1_cfg.snat().external_ip,
+        mapped_port,
+        seq_no,
+        &data[..],
+    );
+
+    assert_eq!(g1.port.stats_snap().in_uft_hit, 0);
+    let res = g1.port.process(In, &mut pkt4, &mut ameta);
+    assert!(matches!(res, Ok(Modified)), "bad result: {:?}", res);
+    assert_port!(g1);
+
+    // Inner Ether + IP
+    assert_eq!(pkt4.body_offset(), 14 + 20);
+    assert_eq!(pkt4.body_seg(), 1);
+    let meta = pkt4.meta();
+
+    match meta.inner.ether.as_ref() {
+        Some(eth) => {
+            assert_eq!(eth.src, g1_cfg.boundary_services.mac);
+            assert_eq!(eth.dst, g1_cfg.private_mac);
+            assert_eq!(eth.ether_type, ETHER_TYPE_IPV4);
+        }
+
+        None => panic!("expected inner Ether header"),
+    }
+
+    match meta.inner.ip.as_ref().unwrap() {
+        IpMeta::Ip4(ip4) => {
+            assert_eq!(ip4.src, dst_ip);
+            assert_eq!(ip4.dst, g1_cfg.ipv4().private_ip);
+            assert_eq!(ip4.proto, Protocol::ICMP);
+        }
+
+        ip6 => panic!("execpted inner IPv4 metadata, got IPv6: {:?}", ip6),
+    }
+
+    let body = pkt4.body_segs().unwrap()[0];
+    let icmp = Icmpv4Packet::new_checked(body).unwrap();
+    assert_eq!(icmp.echo_ident(), ident);
```

Comment:
This is really a question, but I want to make sure I understand what's going on here.

This asserts that the _incoming_ reply has been modified to use the original `ident` from the outbound echo request, right? So the full scenario is:

- Send an ICMP echo request with some ident
- On the way out, that's modified to use one of the mapped ports from the SNAT configuration
- On the way back in, we're looking for that mapped port, and we modify the packet again from the mapped port back to the original `ident`.

Is that all accurate?

Reply:
Yes, the steps you described are exactly the purpose of the SNAT/ICMP body transform. On outbound we need to map the Echo Identifier into the SNAT port space. On inbound we need to do the reverse, and map it back to the Echo Identifier that the guest is expecting to see.